cartography/intel/aws/permission_relationships.py

```diff
@@ -0,0 +1,300 @@
+    """
+    if clause_name not in statement:
+        return missing_clause_return
+    for clause in statement[clause_name]:
+        if evaluate_clause(clause, match):
+            return True
+    return False
+
+
+def evaluate_statements_for_permission(statements, permission, resource_arn):
+    """ Evaluate an entire statement for a specific permission against a resource
+
+    Arguments:
+        statements {[dict]} -- The list of statements to be evaluated
+        permission {str} -- The permission to evaluate
+        resource_arn {[type]} -- The resource to test the permission against
+
+    Returns:
+        [bool] -- If the statement grants the specific permission to the resource
+    """
+    allowed = False
+    for statement in statements:
+        if not evaluate_statement_clause_for_permission(statement, "notaction", permission):
+            if evaluate_statement_clause_for_permission(statement, "action", permission, missing_clause_return=True):
+                if evaluate_statement_clause_for_permission(statement, "resource", resource_arn):
+                    if not evaluate_statement_clause_for_permission(statement, "notresource", resource_arn):
+                        return True
+
+    return allowed
+
+
+def evaluate_policy_for_permission(statements, permissions, resource_arn):
+    """ Evaluates an entire policy for specific permissions to a resource.
+    AWS Policy evaluation reference
+    https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_evaluation-logic.html
+
+    Arguments:
+        statements {[dict]} -- The list of statements for the policy
+        permissions {[str]} -- The permissions to evaluate
+        resource_arn {[type]} -- The resource to test the permission against
+
+    Returns:
+        [(bool, bool)] -- (allowed_by_policy, explicitly_denied_by_policy)
+        return cases
+        (True, False) - The policy allows the action
+        (False, False) - The poliy does not allow the action, but also doesn't explicitly deny it
+        (False, True) - The policy specifically denies the action. There is no need to evaluate other policies
+    """
+    allow_statements = [s for s in statements if s["effect"] == "Allow"]
+    deny_statements = [s for s in statements if s["effect"] == "Deny"]
+    for permission in permissions:
+        if evaluate_statements_for_permission(deny_statements, permission, resource_arn):
+            # The action explicitly denied then no other policy can override it
+            return False, True
+        else:
+            if evaluate_statements_for_permission(allow_statements, permission, resource_arn):
+                # The action is allowed by this policy
+                return True, False
+    # The action is not allowed by this policy, but not specifically denied either
+    return False, False
+
+
+def evaluate_policies_against_resource(policies, resource_arn, permissions):
```

Comment:
Reading 
```
if evaluate_policies_against_resource(policies, resource_arn, permission):
```

could probably be a bit smoother. How about something like `principal_allowed_on_resource()`? (it's pretty late and I'm running out of ideas)

Reply:
done

Reply:
Ignore  this.